File: biosilico-creatures/src/main/resources/creatures/creaturesDocs/geneticsParsing/python/creatures1GenomeParsing.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import sys
import struct

import GeneSwitcherCreatures
from GeneSwitcherCreatures import GeneSwitcherCreatures1
from GeneSwitcherCreatures import GeneEnumGroups

file2parse = None
if (len(sys.argv) < 2) : 
  print( "Need file to parse in argument !" )
  exit( 1 )
else : 
  file2parse = sys.argv[1]

## use struct package : https://docs.python.org/3/library/struct.html
## see also : https://www.tutorialspoint.com/struct-module-in-python

## NOTEs of DEV of this script : 
## ## ## '.gen' files appears somehow corrupted and script here (with module) take account of these ; 
## ## ## 1/ gene longer than expected : just cut off !
## ## ## 2/ gene shorter than expected : completion ?!

def readBinaryGenes(stream) : 
  data = b''
  while True:
    bytes = stream.read(1)
    if bytes == "":
      break;
    data += bytes
    ## NOTE : optimize this below !!
    if (data.endswith( b'gene' )) or (data.endswith( b'gend' )):
      break;
  return data
    
switcher = GeneSwitcherCreatures1()
listOfGenes = []
with open(file2parse, 'rb') as bfile :
  while True : 
    data = readBinaryGenes( bfile )
    ## print( data )
    nextgene = switcher.define4creatures1( data[:-4] )
    ## print( nextgene )
    ## print( type( nextgene ) )
    if ( type(nextgene) is GeneSwitcherCreatures.GeneCreatures1) : 
      listOfGenes.append( nextgene )
    if (data.endswith( b'gend') ) : 
      print( "END OF GENOME" )
      break

for gene in listOfGenes : 
  gene.printInLine()

for tsg in GeneEnumGroups.getEnumsTSG() : 
  print( tsg )

File: biosilico-creatures/src/main/resources/creatures/creaturesDocs/geneticsParsing/perl/creatures1GenomeParsing.pl
#!/usr/bin/perl -w

use strict;

use lib '.';
use GeneEnumsGroup;
use DataLoader;

use CreaturesGene;

my $fileInputPath = $ARGV[0];
if (!$fileInputPath) { die "No FilePath in argument !"; }

## To read *.gen Creatures 1 files
## ## ## "./creatures1GlobalParsing.pl ../../genetics/0WKW.gen"
## ## ## "./creatures1GlobalParsing.pl ../../genetics/8ARU.gen"

## https://www.perl.com/article/how-to-parse-binary-data-with-perl/
## https://www.tutorialspoint.com/perl/perl_unpack.htm
## https://www.tutorialspoint.com/perl/perl_pack.htm
## for beginners in OOP in Perl : https://www.tutorialspoint.com/perl/perl_oo_perl.htm
## see also : https://www.perltutorial.org/perl-oop/

sub readBinaryGenes {
	my $fileReadingInput = shift;
	my $data = "";
	my $loopindex = 1;
	while ( $loopindex ) {
		my $bytes = undef;
		my $success = read $fileReadingInput, $bytes, 1;
		## print $bytes . " // " . unpack("a<*", $data)." // ".pack("a4", "gene")."\n";
		$data .= $bytes;
		## exit(1);
		if ( ($data =~ /(.*?)gen[ed]$/ ) || ( ! $success) ) {
			## print "gene detected !";
			$loopindex = 0;
		}
		## getc();
	}
	return $data;
}

my @listOfGenes = ();

open my $fileReadingInput, "<", $fileInputPath or die "File does not exist !";
binmode $fileReadingInput ;

while( my $gene = readBinaryGenes( $fileReadingInput ) ) {
	if ($gene eq "gene")	{ next; }
	if ($gene eq "")		{ next; }
	if ($gene =~ /gend/)	{ last; }
	$gene =~ s/gen[ed]$//;
	## print $gene."\n";
	my @toTreat = unpack("(C*)<", $gene);
	
	my $nextGene = &CreaturesGene::treatGeneData( @toTreat );
	
	## NOTE : issue with 6JHJ ; 7WEW ; 8PEQ (grendel !) and 1QOI ?!
	## TODO compare perl / python export !!
	
	if (defined $nextGene) { push( @listOfGenes, $nextGene ); }
	
}

close $fileReadingInput;

foreach my $gene (@listOfGenes) {
	print( $gene->toString() );
}

my $genomeName = $fileInputPath;
$genomeName =~ s/.*?\/([0-9A-Z]{4})\.gen$/$1/;
print( "Number of genes (".@listOfGenes.") {". $genomeName."}\n" );

